plugins/CoreHome/javascripts/dataTable.js

```diff
@@ -40,17 +40,34 @@ function DataTable(element) {
 DataTable._footerIconHandlers = {};
 
 DataTable.initNewDataTables = function (reportId) {
-    var selector = typeof reportId === 'string' ? '[data-report='+JSON.stringify(reportId)+']' : 'div.dataTable';
+    let retryInvisible = false;
+
+    const selector = typeof reportId === 'string' ? '[data-report='+JSON.stringify(reportId)+']' : 'div.dataTable';
+
     $(selector).each(function () {
-        if (!$(this).attr('id')) {
-            var tableType = $(this).attr('data-table-type') || 'DataTable',
-                klass = require('piwik/UI')[tableType] || require(tableType);
+        if ($(this).attr('id')) {
+            return;
+        }
 
-            if (klass && $.isFunction(klass)) {
-                var table = new klass(this);
-            }
+        if (!$(this).is(':visible')) {
+            retryInvisible = true;
+
+            return;
+        }
+
+        const tableType = $(this).attr('data-table-type') || 'DataTable';
+        const klass = require('piwik/UI')[tableType] || require(tableType);
+
+        if (klass && $.isFunction(klass)) {
+            new klass(this);
         }
     });
+
+    if (retryInvisible) {
+        setTimeout(function() {
+            DataTable.initNewDataTables(reportId);
+        }, 50);
+    }
```

Comment:
Was wondering if this could end up in some sort of endless loop if the report doesn't get visible at all. In that case it would retry over and over again. Should we maybe pass some sort of counter and stop after 3 or 5 retries and maybe increase the timeout with each retry or similar?

Reply:
And if we run out of tries/out of time, should we also display an error message? Otherwise it may end up in a similar blank widget like how it is now.

Reply:
The code would retry until either the element was displayed or removed from the DOM. In that way it could potentially loop endlessly. Limiting the retries would obviously prevent that, but we wouldn't know if we had enough retries. That also makes it nigh impossible to display any meaningful error message.

So, how about a different approach?

I changed the `Widget` to only start running the `WidgetLoader` if the middleware says it will be displayed. This forces the requests into the order required to prevent the invislble evolution graph, but I guess the delay from the now sequential middleware request should be reasonably low to ignore the sequential execution. If middleware parameters are involved, this prevents any request for a widget that won't get shown.

Reply:
Sounds like a reasonable approach to me. There shouldn't be any sense in already starting to request data in a widget without knowing if it should be displayed or not.